### PR TITLE
Export default configurations for grafana and prometheus

### PR DESCRIPTION
**Pull request by alanwguo**
Users who are using prometheus and grafana and use these default configurations with to show good default visualizations with those tools

Signed-off-by: Alan Guo <aguo@anyscale.com>

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: dashboard/modules/metrics/metrics_head.py
Comment (rickyyx) [inline dashboard/modules/metrics/metrics_head.py:32]: Any impact/need of making this a non-minimal module?

File: dashboard/modules/metrics/prometheus/prometheus.yml
Comment (rickyyx) [inline dashboard/modules/metrics/prometheus/prometheus.yml:8]: looks like this is just default?

File: dashboard/modules/metrics/prometheus/prometheus.yml
Comment (rickyyx) [inline dashboard/modules/metrics/prometheus/prometheus.yml:16]: do we need these settings?

File: dashboard/modules/metrics/metrics_head.py
Comment (rickyyx) [inline dashboard/modules/metrics/metrics_head.py:64]: nit: could we log both the prometheus and grafana config output paths maybe?

**Review by rickyyx — COMMENTED** [review]
LGTM! Left a few questions and small nits. 

Also, 
1. Is this something we need to document publicly somewhere else? 
2. Do we need test (manual/e2e automated) for a new head module in general? cc @rkooo567

File: dashboard/modules/metrics/metrics_head.py
Comment (galenhwang) [inline dashboard/modules/metrics/metrics_head.py:46]: nit: extract datasources path to variable, affects L50-L55

File: dashboard/modules/metrics/prometheus/prometheus.yml
Comment (galenhwang) [inline dashboard/modules/metrics/prometheus/prometheus.yml:19]: nit: Fix/remove comment

**Review by rkooo567 — COMMENTED** [review]
Is the goal of metricsHead module copying&pasting configs? Maybe we should make this optional. Also we need a documentation to add this feature (and maybe an API review).

**Comment by alanwguo** [comment]
@rkooo567 @galenhwang @rickyyx , for more context: here is the e2e prototype PR: https://github.com/ray-project/ray/pull/28285

metrics-head ends up doing a bit more such as detecting if grafana is running healthily. 
sceenshot of the final dashboard is in that PR.

**Comment by alanwguo** [comment]
> LGTM! Left a few questions and small nits.
> 
> Also,
> 
> 1. Is this something we need to document publicly somewhere else?
> 2. Do we need test (manual/e2e automated) for a new head module in general? cc @rkooo567

Yes, I'll write both documentation and a test for this.

I'm thinking the docs will be something like this:


1. Install prom / grafana yourself.
2. Launch ray with `RAY_PROMETHEUS_HOST` override if necessary.
3. Launch prom and grafana with `--config` parameters that point to the proper config in `/tmp/ray/metrics`.
4. Visit grafana to see the default dashboard.

File: dashboard/modules/metrics/metrics_head.py
Comment (alanwguo) [reply dashboard/modules/metrics/metrics_head.py:18]: Good question. I was wondering the same thing. Today `/tmp/ray/prom_metrics_service_discovery.json` is not session specific

File: dashboard/modules/metrics/grafana/grafana.ini
Comment (alanwguo) [reply dashboard/modules/metrics/grafana/grafana.ini:7]: what do you mean by that? 

If you mean allow users to change settings, we can add in the documentation instructions on how to override the settings.

File: dashboard/modules/metrics/grafana/provisioning/dashboards/default_grafana_dashboard.json
Comment (alanwguo) [reply dashboard/modules/metrics/grafana/provisioning/dashboards/default_grafana_dashboard.json:1]: > @rkooo567 @galenhwang @rickyyx , for more context: here is the e2e prototype PR: #28285
> 
> metrics-head ends up doing a bit more such as detecting if grafana is running healthily. sceenshot of the final dashboard is in that PR.

See the PR form this comment.

File: dashboard/modules/metrics/grafana/grafana.ini
Comment (galenhwang) [reply dashboard/modules/metrics/grafana/grafana.ini:7]: That makes sense. For default though, should we consider setting `org_role = Editor`?

File: dashboard/modules/metrics/grafana/provisioning/dashboards/default_grafana_dashboard.json
Comment (galenhwang) [reply dashboard/modules/metrics/grafana/provisioning/dashboards/default_grafana_dashboard.json:1]: What does the metrics tab show if Prometheus / Grafana aren't installed and running?

Comment (alanwguo) [reply dashboard/modules/metrics/grafana/provisioning/dashboards/default_grafana_dashboard.json:1]: We will show an error message saying to please install prometheus / grafana if they want to see time series metrics

File: dashboard/modules/metrics/metrics_head.py
Comment (rkooo567) [inline dashboard/modules/metrics/metrics_head.py:34]: Can you add a test this code is not broken when it is run twice? (e.g., when a new ray instance is started)

File: dashboard/modules/metrics/metrics_head.py
Comment (rkooo567) [inline dashboard/modules/metrics/metrics_head.py:37]: What happens if users provide their own config?

Comment (rkooo567) [reply dashboard/modules/metrics/metrics_head.py:37]: Maybe we need an API for users to provide their own config (e.g., add a template to a specific folder).

Comment (alanwguo) [reply dashboard/modules/metrics/metrics_head.py:37]: I'll create a folder `/tmp/ray/metrics/grafana/custom-dashboards`. Any dashboard json found there can be automatically picked up. We won't ever delete that folder.

Comment (rkooo567) [reply dashboard/modules/metrics/metrics_head.py:37]: Is it going to be in this PR?

Comment (alanwguo) [reply dashboard/modules/metrics/metrics_head.py:37]: yes.. in progress

**Review by DmitriGekhtman — APPROVED** [review]
cluster doc codeowner signoff

File: dashboard/modules/metrics/metrics_head.py
Comment (rkooo567) [inline dashboard/modules/metrics/metrics_head.py:16]: Should this folder be configurable?

File: doc/source/ray-observability/ray-metrics.rst
Comment (rkooo567) [inline doc/source/ray-observability/ray-metrics.rst:44]: Can you mention this is available only in `ray[default]`?

File: python/ray/tests/test_metrics_head.py
Comment (rkooo567) [inline python/ray/tests/test_metrics_head.py:18]: Should we test these are not created when `include_dashboard = False`?

File: doc/source/ray-observability/ray-metrics.rst
Comment (rkooo567) [inline doc/source/ray-observability/ray-metrics.rst:94]: I think we also need a section for custom prometheus & Grafana config

File: dashboard/modules/metrics/metrics_head.py
Comment (rkooo567) [inline dashboard/modules/metrics/metrics_head.py:89]: I think we need a custom Prometheus folder as well (e.g., some users may want to tweak prometheus config). Could be done as a follow up though

File: python/ray/tests/test_metrics_head.py
Comment (rkooo567) [inline python/ray/tests/test_metrics_head.py:39]: Can you do this inside the fixture. Otherwise it might be flaky

File: python/ray/tests/test_metrics_head.py
Comment (rkooo567) [inline python/ray/tests/test_metrics_head.py:13]: I think we should add test_metrics_head.py to python/ray/tests/BUILD

**Review by rkooo567 — APPROVED** [review]
I think we should discuss what's the right API for the custom Grafana dashboard. But I feel like this could be in a separate PR (and we can only focus on the default one). So let's defer this for now? Other things lgtm if last comments are addressed

**Comment by rkooo567** [comment]
Also tests seem to fail now

File: doc/source/ray-observability/ray-metrics.rst
Comment (richardliaw) [inline doc/source/ray-observability/ray-metrics.rst:74]: dont forget to remove this :)

File: doc/source/ray-observability/ray-metrics.rst
Comment (richardliaw) [reply doc/source/ray-observability/ray-metrics.rst:44]: do we also record this per ray session? ie why not scope it to 'tmp/ray/session_latest"?

File: doc/source/cluster/running-applications/monitoring-and-observability.rst
Comment (richardliaw) [inline doc/source/cluster/running-applications/monitoring-and-observability.rst:113]: can you link to Prometheus config documentation here?

**Review by richardliaw — COMMENTED** [review]
nice! ping when you need a final review?

File: doc/source/ray-observability/ray-metrics.rst
Comment (alanwguo) [reply doc/source/ray-observability/ray-metrics.rst:44]: yes, changed to be session scoped. Will need to update `prom_metrics_service_discovery.json` as well in a future PR.

File: doc/source/ray-observability/ray-metrics.rst
Comment (alanwguo) [reply doc/source/ray-observability/ray-metrics.rst:94]: removed custom configs for now.

File: dashboard/modules/metrics/export/prometheus/prometheus.yml
Comment (ericl) [inline dashboard/modules/metrics/export/prometheus/prometheus.yml:12]: This is pretty cool.

**Review by ericl — APPROVED** [review]

**Pull request merged**
